File: internal/bridge/protocol.go
// Package bridge carries control messages and media frames between the
// WebView frontend and the Go backend over a loopback WebSocket.
//
// The WebView owns capture and playback: it holds the camera and
// microphone, runs the WebCodecs encoders and decoders, and paints the
// video grid. The Go side owns the MOQ session. Everything in between
// travels over one WebSocket connection: control state as JSON text
// frames, encoded media as binary frames with the compact header defined
// below.
//
// A WebSocket rather than Wails bindings because media is binary and hot:
// bindings marshal through JSON (so every frame would pay base64), while
// binary WebSocket frames on loopback measure ~400 MB/s up and ~1 GB/s
// down in this WebView — three orders of magnitude more than a conference
// needs.
package bridge

import "encoding/json"

// Media frame header. Fixed 24 bytes, big-endian, followed by an optional
// codec-config blob and then the encoded frame payload:
//
//	 0      u8      Version (always FrameVersion)
//	 1      u8      Kind (KindVideo / KindAudio)
//	 2      u8      Flags (FlagKeyFrame, FlagAudioLevel, TemporalLayer)
//	 3      u8      AudioLevel — valid only with FlagAudioLevel
//	 4..8   u32     Handle — identifies the track (see below)
//	 8..16  u64     Timestamp in microseconds, from the encoder
//	16..20  u32     ConfigLen — length of the config blob that follows
//	20..24  u32     PayloadLen — length of the payload after the config
//
// Handle namespaces differ by direction. Frontend to backend, it names
// one of the two local publications (HandleLocalVideo / HandleLocalAudio).
// Backend to frontend, it is a backend-assigned per-remote-track handle,
// announced ahead of any frame by a "track" control message.
const (
	FrameVersion   = 1
	FrameHeaderLen = 24

	KindVideo = 0
	KindAudio = 1

	FlagKeyFrame = 1 << 0
	// FlagAudioLevel marks the AudioLevel byte as carrying a real
	// measurement. Needed because level 0 means "loudest", not "absent".
	FlagAudioLevel = 1 << 1

	// TemporalLayerShift and TemporalLayerMask carve the frame's temporal
	// layer out of the flags byte: bits 2-3, so layers 0-3.
	//
	// In the flags byte rather than a field of its own because the header is
	// a fixed 24 bytes with two hand-written implementations, and widening it
	// would be a wire break for a value that needs two bits. Byte 3 was the
	// other candidate and is worse: it belongs to audio, and overloading it
	// per-kind is how the two implementations drift.
	//
	// Zero is the base layer, which is also what a frame from an encoder
	// producing no temporal layers reports — so an unlayered stream is layer
	// 0 throughout and needs no special case anywhere.
	TemporalLayerShift = 2
	TemporalLayerMask  = 0x3 << TemporalLayerShift

	// MaxTemporalLayer is the highest layer the two bits above can carry.
	MaxTemporalLayer = 3
)

// Handles for the two tracks the frontend publishes. Remote handles are
// assigned from HandleRemoteBase upward so they can never collide.
const (
	HandleLocalVideo = 0
	HandleLocalAudio = 1
	// HandleLocalVideoLow is the second, smaller encoding of the same camera —
	// see KindVideoLow. Its frames are KindVideo like any other picture; only
	// the handle says which encoding they belong to.
	HandleLocalVideoLow = 2
	HandleRemoteBase    = 16
)

// MediaFrame is one encoded chunk in flight over the bridge.
type MediaFrame struct {
	Kind      uint8
	Handle    uint32
	Timestamp uint64
	KeyFrame  bool
	// Config is the codec description (WebCodecs
	// VideoDecoderConfig.description / AudioDecoderConfig.description).
	// Present only on frames where the encoder emitted a new config —
	// which for Opus is the first frame, and for H.264 in Annex B is
	// never, since SPS/PPS travel in-band.
	Config  []byte
	Payload []byte

	// AudioLevel is the RFC 6464 byte LOC's AudioLevel property carries
	// (§2.3.3.2): bit 7 is voice activity, bits 0-6 the magnitude in
	// -dBov, where 0 is loudest and 127 is silence. Valid only when
	// HasAudioLevel is set. It drives the speaking indicator, so it
	// travels with every audio frame in both directions.
	AudioLevel    uint8
	HasAudioLevel bool

	// TemporalLayer is which temporal layer of an SVC encoding this frame
	// belongs to: 0 is the base, which every higher layer references and which
	// decodes on its own. It decides the subgroup the publisher writes the
	// frame to, so a subscriber or relay can shed the upper layers without
	// touching the base. Always 0 for audio, and for video from an encoder
	// configured without temporal layers.
	TemporalLayer uint8
}

// ---- control messages: frontend to backend ----------------------------

// ClientMessage is a JSON control frame from the WebView.
type ClientMessage struct {
	Type string `json:"type"`

	Join     *JoinRequest  `json:"join,omitempty"`
	Track    *TrackConfig  `json:"track,omitempty"`
	Stats    *ClientStats  `json:"stats,omitempty"`
	LogLevel string        `json:"logLevel,omitempty"`
	Report   *ClientReport `json:"report,omitempty"`
	// Untrack names a kind ("video"/"audio") the frontend has stopped
	// publishing, so the catalog stops advertising it.
	Untrack string `json:"untrack,omitempty"`
	// OpenURL is a link to hand to the OS browser. See MsgOpenURL.
	OpenURL string `json:"openUrl,omitempty"`
	// Interest is which remote video is worth receiving. See MsgInterest.
	Interest *Interest `json:"interest,omitempty"`
}

// Interest is the frontend's account of which remote participants' video it
// can currently make use of: the tiles on screen, plus whatever a scroll is
// about to bring on.
//
// Only video. Audio is never gated on being visible — someone speaking
// off-screen has to be heard, and the speaking indicator is what makes anyone
// scroll to them in the first place. It is also 32 kbps against video's
// 1.5 Mbps, so there is nothing to win by dropping it.
type Interest struct {
	// Video lists participant IDs whose video is wanted. An empty list means
	// none, which is a real answer — everyone can be scrolled off screen.
	Video []string `json:"video"`
	// Low is the subset of Video whose tile is small enough that the
	// publisher's smaller encoding will do. Anything here that is not also in
	// Video is not subscribed at all: this says which encoding, never whether.
	Low []string `json:"low,omitempty"`
}

// Client message types.
const (
	MsgJoin     = "join"     // dial the relay and start publishing
	MsgLeave    = "leave"    // tear the session down
	MsgTrack    = "track"    // declare the codec config for a local track
	MsgStats    = "stats"    // frontend-side encode/decode counters
	MsgLogLevel = "logLevel" // change the backend slog level
	MsgReport   = "report"   // a frontend-side event worth logging
	// MsgUntrack withdraws a local track. Turning the camera off stops the
	// frames, but a catalog that still declares video leaves every
	// subscriber holding a decoder and showing its last frame forever — so
	// stopping has to be said, not merely done.
	MsgUntrack = "untrack"
	// MsgOpenURL hands a link to the OS. The WebView must not navigate to it
	// itself — that would replace the app with a web page and there would be
	// no way back — and a target=_blank anchor generally does nothing here.
	MsgOpenURL = "openUrl"

	// MsgInterest carries which remote participants' video is worth
	// receiving — the tiles on screen, plus a margin for scrolling.
	MsgInterest = "interest"
)

// ClientReport is something the frontend wants in the shared log: a
// capture failure, a decoder error, a lifecycle event. Without this the
// two halves keep separate logs and the debug panel only shows half the
// story.
type ClientReport struct {
	Level   string            `json:"level"`
	Message string            `json:"msg"`
	Attrs   map[string]string `json:"attrs,omitempty"`
}

// JoinRequest asks the backend to dial a relay and join a room.
type JoinRequest struct {
	Relay    string `json:"relay"`
	Room     string `json:"room"`
	Nickname string `json:"nickname"`
}

// TrackConfig describes one local publication well enough for a remote
// subscriber to configure a matching decoder. It is what the backend
// writes into its MSF catalog, and what the frontend sends once its
// encoder has produced a decoder config.
type TrackConfig struct {
	Kind string `json:"kind"` // "video" | "audio"

	Codec string `json:"codec"` // WebCodecs codec string

	// Description is the codec extradata, base64-encoded — WebCodecs
	// VideoDecoderConfig/AudioDecoderConfig `description`. Empty for
	// Annex B H.264.
	Description string `json:"description,omitempty"`

	Width     uint32  `json:"width,omitempty"`
	Height    uint32  `json:"height,omitempty"`
	Framerate float64 `json:"framerate,omitempty"`
	Bitrate   uint64  `json:"bitrate,omitempty"`

	SampleRate uint32 `json:"sampleRate,omitempty"`
	Channels   uint32 `json:"channels,omitempty"`

	// TemporalLayers is how many temporal layers the encoder was configured
	// to emit, and so how many subgroups a group of this track is published
	// on. One or zero both mean a single subgroup, which is what every track
	// was before layers and what an encoder that ignores scalabilityMode
	// produces anyway.
	//
	// A subscriber needs this before the first frame, not after: a group's
	// subgroup streams do not arrive together, and reassembly cannot tell a
	// layer that is gone from one the relay has not drained yet unless it
	// knows how many to look for. See internal/conf/reorder.go.
	TemporalLayers uint8 `json:"temporalLayers,omitempty"`
}

// ClientStats are the frontend's own counters, forwarded so the debug
// panels can show capture and playback health next to transport health.
type ClientStats struct {
	EncodeFPS      float64            `json:"encodeFps"`
	EncodeQueue    int                `json:"encodeQueue"`
	EncodeKbps     float64            `json:"encodeKbps"`
	AudioEncodeFPS float64            `json:"audioEncodeFps"`
	Decoders       map[string]float64 `json:"decoders,omitempty"`
}

// ---- control messages: backend to frontend ----------------------------

// ServerMessage is a JSON control frame to the WebView.
type ServerMessage struct {
	Type string `json:"type"`

	State        *SessionState  `json:"state,omitempty"`
	Participants []Participant  `json:"participants,omitempty"`
	Track        *RemoteTrack   `json:"track,omitempty"`
	TrackGone    *RemoteTrackID `json:"trackGone,omitempty"`
	Log          *LogEntry      `json:"log,omitempty"`
	Metrics      *Metrics       `json:"metrics,omitempty"`
	Invite       *Invite        `json:"invite,omitempty"`
	Update       *Update        `json:"update,omitempty"`
	Error        string         `json:"error,omitempty"`
}

// Update is a released version newer than the one running. Sent at most once
// per run, and only when there is something to offer — the frontend shows a
// button when it arrives and nothing at all when it does not.
type Update struct {
	Version string `json:"version"`
	URL     string `json:"url"`
}

// Invite is a relay and room the app was asked to join from outside — an
// invite link the OS handed us. The frontend fills its welcome form from it
// and joins.
type Invite struct {
	Relay string `json:"relay"`
	Room  string `json:"room"`
}

// Server message types.
const (
	MsgState        = "state"
	MsgParticipants = "participants"
	MsgRemoteTrack  = "remoteTrack"
	MsgTrackGone    = "trackGone"
	MsgLog          = "log"
	MsgMetrics      = "metrics"
	MsgInvite       = "invite"
	MsgUpdate       = "update"
	MsgError        = "error"
	// MsgRequestKeyFrame asks the frontend's video encoder for an immediate
	// keyframe. A fresh session has no open group, and writeVideo will not
	// start one on a delta frame, so without this a reconnect stays blank
	// until the next scheduled keyframe — up to the keyframe interval.
	MsgRequestKeyFrame = "requestKeyFrame"
)

// Session lifecycle states reported in SessionState.Phase.
const (
	PhaseIdle       = "idle"
	PhaseConnecting = "connecting"
	PhaseJoined     = "joined"
	PhaseFailed     = "failed"
	// PhaseReconnecting means the room was joined and the relay session
	// then ended; the backend is re-dialling. Distinct from connecting so
	// the frontend can keep the call on screen and say what is happening
	// rather than dumping the user back to the welcome form.
	PhaseReconnecting = "reconnecting"
)

// SessionState is the backend's view of the MOQ session.
type SessionState struct {
	Phase string `json:"phase"`
	Relay string `json:"relay,omitempty"`
	Room  string `json:"room,omitempty"`
	// ID is the local participant's generated identifier — the last
	// field of its namespace tuple.
	ID       string `json:"id,omitempty"`
	Nickname string `json:"nickname,omitempty"`
	Detail   string `json:"detail,omitempty"`
}

// Participant is one remote peer discovered in the room.
type Participant struct {
	ID       string `json:"id"`
	Nickname string `json:"nickname"`
	// Version is the build they are running, or empty from a peer old
	// enough not to publish one.
	Version  string `json:"version,omitempty"`
	HasVideo bool   `json:"hasVideo"`
	HasAudio bool   `json:"hasAudio"`
	// VideoLevel is how much of their video this client is still taking:
	// "full", "small", or "none". Distinct from HasVideo, which says what they
	// publish — a peer we have given up on and a peer with the camera off both
	// show no picture, and only this tells them apart.
	VideoLevel string `json:"videoLevel,omitempty"`
}

// RemoteTrack announces an inbound track and the handle its frames will
// carry. The frontend configures a decoder from Config and then renders
// every binary frame bearing Handle.
type RemoteTrack struct {
	Handle      uint32      `json:"handle"`
	Participant string      `json:"participant"`
	Nickname    string      `json:"nickname"`
	Config      TrackConfig `json:"config"`
}

// RemoteTrackID identifies a track that has ended.
type RemoteTrackID struct {
	Handle      uint32 `json:"handle"`
	Participant string `json:"participant"`
}

// LogEntry is one backend log record, forwarded to the debug log panel.
type LogEntry struct {
	// TimeMillis is a Unix millisecond timestamp.
	TimeMillis int64             `json:"t"`
	Level      string            `json:"level"`
	Message    string            `json:"msg"`
	Attrs      map[string]string `json:"attrs,omitempty"`
}

// Metrics is one sample of transport and per-track health, emitted on a
// fixed interval so the frontend can plot it directly.
type Metrics struct {
	TimeMillis int64 `json:"t"`

	// QUIC transport, from the connection's qlog event stream.
	SmoothedRTTMillis float64 `json:"rttMs"`
	MinRTTMillis      float64 `json:"minRttMs"`
	LatestRTTMillis   float64 `json:"latestRttMs"`
	// PeakRTTMillis is the worst RTT measured during the sample interval —
	// the spike a smoothed average hides.
	PeakRTTMillis    float64 `json:"peakRttMs"`
	CongestionWindow int     `json:"cwnd"`
	BytesInFlight    int     `json:"bytesInFlight"`
	PacketsInFlight  int     `json:"packetsInFlight"`
	CongestionState  string  `json:"congestionState,omitempty"`

	// Rates over the sample interval.
	PacketsSentPerSec float64 `json:"packetsSentPerSec"`
	PacketsLostPerSec float64 `json:"packetsLostPerSec"`
	LossPercent       float64 `json:"lossPercent"`
	SendKbps          float64 `json:"sendKbps"`
	ReceiveKbps       float64 `json:"receiveKbps"`

	// Cumulative counters.
	PacketsSent     uint64 `json:"packetsSent"`
	PacketsReceived uint64 `json:"packetsReceived"`
	PacketsLost     uint64 `json:"packetsLost"`

	// MOQ-level rates, split by direction.
	PublishKbps   float64 `json:"publishKbps"`
	SubscribeKbps float64 `json:"subscribeKbps"`
	// Objects and groups published/received over the interval.
	ObjectsOutPerSec float64 `json:"objectsOutPerSec"`
	ObjectsInPerSec  float64 `json:"objectsInPerSec"`
	GroupsOutPerSec  float64 `json:"groupsOutPerSec"`

	// BridgeDropped is how many frames this bridge has dropped on its way to
	// the WebView, cumulative for the connection. Loss here is not the
	// network's and does not show up anywhere else: the send queue is bounded
	// and a full one discards rather than blocks, so a WebView that cannot
	// drain fast enough loses audio and video with nothing else to say so.
	BridgeDropped uint64 `json:"bridgeDropped"`

	Tracks []TrackMetrics `json:"tracks,omitempty"`
}

// TrackMetrics is per-track throughput, keyed by a label the frontend can
// show verbatim (e.g. "out/video" or "in/ab12cd/audio").
type TrackMetrics struct {
	Label   string  `json:"label"`
	Kbps    float64 `json:"kbps"`
	Objects uint64  `json:"objects"`
	Groups  uint64  `json:"groups"`
	// SkewMillisPerSec is how fast this track's arrivals are falling behind
	// the clock that produced them, in milliseconds per second. Positive means
	// a queue on the inbound path is filling — the path is carrying less than
	// is being sent. Inbound audio only, and nil until there is enough history
	// to fit a trend, which is why it is a pointer: zero means "keeping up",
	// not "no reading".
	SkewMillisPerSec *float64 `json:"skewMillisPerSec,omitempty"`
	// LagMillis is how much later objects are arriving now than when this
	// subscription started — the integral of the above. It says the call has
	// fallen behind, where the slope says it is falling behind.
	LagMillis *float64 `json:"lagMillis,omitempty"`
}

// Endpoint is what the asset handler serves at /__bridge so the frontend
// can find and authenticate to the WebSocket.
type Endpoint struct {
	URL   string `json:"url"`
	Token string `json:"token"`
	// Version is the build the backend is running, which is also the build
	// the frontend is part of — they ship in one binary. Carried here rather
	// than in a control message because the welcome screen wants to show it
	// before any session exists, and this descriptor is already the first
	// thing the frontend fetches.
	Version string `json:"version,omitempty"`
}

func (e Endpoint) JSON() []byte {
	b, _ := json.Marshal(e)
	return b
}

File: frontend/src/lib/protocol.ts
/**
 * Wire types and framing shared with the Go backend. Keep in step with
 * internal/bridge/protocol.go and internal/bridge/frame.go.
 */

export const FRAME_VERSION = 1;
export const FRAME_HEADER_LEN = 24;

export const KIND_VIDEO = 0;
export const KIND_AUDIO = 1;

export const FLAG_KEYFRAME = 1 << 0;
/** Marks the header's audio-level byte as a real measurement. */
export const FLAG_AUDIO_LEVEL = 1 << 1;

/**
 * The frame's temporal layer lives in bits 2-3 of the flags byte, so layers
 * 0-3. It rides in the flags rather than a field of its own because the header
 * is a fixed 24 bytes with a hand-written implementation at each end, and
 * widening it would be a wire break for a two-bit value.
 *
 * Layer 0 is the base — decodable alone, referenced by every layer above it —
 * and is also what an encoder producing no temporal layers reports, so an
 * unlayered stream needs no special case.
 */
export const TEMPORAL_LAYER_SHIFT = 2;
export const TEMPORAL_LAYER_MASK = 0x3 << TEMPORAL_LAYER_SHIFT;

/** Handles for the two tracks this frontend publishes. */
export const HANDLE_LOCAL_VIDEO = 0;
export const HANDLE_LOCAL_AUDIO = 1;
/**
 * The second, smaller encoding of the same camera. Its frames are KIND_VIDEO
 * like any other picture — only the handle says which encoding they belong to,
 * and only the publisher ever distinguishes them. A subscriber takes one layer
 * or the other and is told about it as plain video.
 */
export const HANDLE_LOCAL_VIDEO_LOW = 2;

export interface MediaFrame {
  kind: number;
  handle: number;
  /** Microseconds, straight from the encoder. */
  timestamp: number;
  keyFrame: boolean;
  /** Codec description bytes, when the encoder emitted a new config. */
  config?: Uint8Array;
  payload: Uint8Array;
  /**
   * RFC 6464 byte from LOC's AudioLevel property: bit 7 voice activity,
   * bits 0-6 magnitude in -dBov. Audio frames only.
   */
  audioLevel?: number;
  /**
   * Which temporal layer of an SVC encoding this frame belongs to. Decides the
   * subgroup the publisher writes it to, so the upper layers can be shed
   * without disturbing the base. Absent means the base layer.
   */
  temporalLayer?: number;
}

/** Encodes one media frame into the binary layout the backend parses. */
export function encodeFrame(f: MediaFrame): ArrayBuffer {
  const configLen = f.config ? f.config.byteLength : 0;
  const buf = new ArrayBuffer(FRAME_HEADER_LEN + configLen + f.payload.byteLength);
  const view = new DataView(buf);
  const bytes = new Uint8Array(buf);

  view.setUint8(0, FRAME_VERSION);
  view.setUint8(1, f.kind);
  let flags = f.keyFrame ? FLAG_KEYFRAME : 0;
  if (f.audioLevel !== undefined) {
    flags |= FLAG_AUDIO_LEVEL;
    view.setUint8(3, f.audioLevel & 0xff);
  }
  flags |= ((f.temporalLayer ?? 0) << TEMPORAL_LAYER_SHIFT) & TEMPORAL_LAYER_MASK;
  view.setUint8(2, flags);
  view.setUint32(4, f.handle);
  // Timestamps are microseconds in a u64. Number stays exact to 2^53 µs,
  // which is ~285 years — far beyond any session.
  view.setBigUint64(8, BigInt(Math.max(0, Math.round(f.timestamp))));
  view.setUint32(16, configLen);
  view.setUint32(20, f.payload.byteLength);

  if (f.config) {
    bytes.set(f.config, FRAME_HEADER_LEN);
  }
  bytes.set(f.payload, FRAME_HEADER_LEN + configLen);
  return buf;
}

/** Decodes one binary frame received from the backend. */
export function decodeFrame(buf: ArrayBuffer): MediaFrame | null {
  if (buf.byteLength < FRAME_HEADER_LEN) return null;
  const view = new DataView(buf);
  if (view.getUint8(0) !== FRAME_VERSION) return null;

  const configLen = view.getUint32(16);
  const payloadLen = view.getUint32(20);
  if (FRAME_HEADER_LEN + configLen + payloadLen > buf.byteLength) return null;

  const configEnd = FRAME_HEADER_LEN + configLen;
  const flags = view.getUint8(2);
  return {
    kind: view.getUint8(1),
    handle: view.getUint32(4),
    timestamp: Number(view.getBigUint64(8)),
    keyFrame: (flags & FLAG_KEYFRAME) !== 0,
    config: configLen > 0 ? new Uint8Array(buf, FRAME_HEADER_LEN, configLen) : undefined,
    payload: new Uint8Array(buf, configEnd, payloadLen),
    audioLevel: (flags & FLAG_AUDIO_LEVEL) !== 0 ? view.getUint8(3) : undefined,
    temporalLayer: (flags & TEMPORAL_LAYER_MASK) >> TEMPORAL_LAYER_SHIFT,
  };
}

// ---- control messages -------------------------------------------------

export interface JoinRequest {
  relay: string;
  room: string;
  nickname: string;
}

export interface TrackConfig {
  kind: 'video' | 'audio';
  codec: string;
  /** Codec extradata, base64. Empty for Annex B H.264. */
  description?: string;
  width?: number;
  height?: number;
  framerate?: number;
  bitrate?: number;
  sampleRate?: number;
  channels?: number;
  /**
   * How many temporal layers the encoder was configured to emit, and so how
   * many subgroups a group of this track is published on. One or absent both
   * mean a single subgroup. A subscriber needs it before the first frame —
   * see internal/conf/reorder.go.
   */
  temporalLayers?: number;
}

export interface ClientStats {
  encodeFps: number;
  encodeQueue: number;
  encodeKbps: number;
  audioEncodeFps: number;
  decoders?: Record<string, number>;
}

/**
 * Something the WebView wants in the backend's log — a capture failure, a
 * decoder error. Without this the two halves keep separate logs and the
 * debug panel only shows half the story.
 */
export interface ClientReport {
  level: 'DEBUG' | 'INFO' | 'WARN' | 'ERROR';
  msg: string;
  attrs?: Record<string, string>;
}

export type ClientMessage =
  | { type: 'join'; join: JoinRequest }
  | { type: 'leave' }
  | { type: 'track'; track: TrackConfig }
  | { type: 'stats'; stats: ClientStats }
  | { type: 'logLevel'; logLevel: string }
  | { type: 'untrack'; untrack: 'video' | 'audio' }
  // Handing a link to the OS is the backend's job: navigating this WebView to
  // a web page would replace the app with it, with no way back to the call.
  | { type: 'openUrl'; openUrl: string }
  // Which remote participants' video is worth receiving. Video only: audio is
  // never gated on being visible, since someone speaking off-screen has to be
  // heard, and hearing them is what makes anyone scroll to them.
  | { type: 'interest'; interest: { video: string[] } }
  | { type: 'report'; report: ClientReport };

/**
 * `reconnecting` means the room was joined and the relay session then ended;
 * the backend is re-dialling. Distinct from `connecting` so the call can stay
 * on screen instead of dropping the user back to the welcome form.
 */
export type Phase = 'idle' | 'connecting' | 'joined' | 'failed' | 'reconnecting';

export interface SessionState {
  phase: Phase;
  relay?: string;
  room?: string;
  id?: string;
  nickname?: string;
  detail?: string;
}

export interface Participant {
  id: string;
  nickname: string;
  /** The build they are running, absent from a peer old enough not to say. */
  version?: string;
  hasVideo: boolean;
  hasAudio: boolean;
  /**
   * How much of their video we are still taking: "full", "small" or "none".
   *
   * Distinct from hasVideo, which says what they publish. A peer the backend
   * gave up on and a peer with the camera off both show no picture, and only
   * this tells them apart.
   */
  videoLevel?: string;
}

/** A released version newer than the one running. */
export interface UpdateInfo {
  version: string;
  url: string;
}

export interface RemoteTrack {
  handle: number;
  participant: string;
  nickname: string;
  config: TrackConfig;
}

export interface RemoteTrackID {
  handle: number;
  participant: string;
}

export interface LogEntry {
  /** Unix milliseconds. */
  t: number;
  level: string;
  msg: string;
  attrs?: Record<string, string>;
}

export interface TrackMetrics {
  label: string;
  kbps: number;
  objects: number;
  groups: number;
  /**
   * How fast this track's arrivals are falling behind the clock that produced
   * them, in milliseconds per second. Positive means a queue on the inbound
   * path is filling — more is being sent to us than is getting through.
   *
   * Inbound audio only, and absent until there is enough history to fit a
   * trend. Absent and zero are different answers: zero means keeping up.
   */
  skewMillisPerSec?: number;
  /**
   * How much later objects are arriving now than when this subscription
   * started — the integral of the above. The slope says the call is falling
   * behind; this says it has fallen behind, and is what a resubscribe exists
   * to reset.
   */
  lagMillis?: number;
}

export interface Metrics {
  t: number;

  rttMs: number;
  minRttMs: number;
  latestRttMs: number;
  peakRttMs: number;
  cwnd: number;
  bytesInFlight: number;
  packetsInFlight: number;
  congestionState?: string;

  packetsSentPerSec: number;
  packetsLostPerSec: number;
  lossPercent: number;
  sendKbps: number;
  receiveKbps: number;

  packetsSent: number;
  packetsReceived: number;
  packetsLost: number;

  publishKbps: number;
  subscribeKbps: number;
  objectsOutPerSec: number;
  objectsInPerSec: number;
  groupsOutPerSec: number;

  /** Frames the bridge dropped on the way here, cumulative for the connection. */
  bridgeDropped: number;

  tracks?: TrackMetrics[];
}

/** A relay and room the OS handed us through an invite link. */
export interface InviteMessage {
  relay: string;
  room: string;
}

export type ServerMessage =
  | { type: 'state'; state: SessionState }
  | { type: 'invite'; invite: InviteMessage }
  | { type: 'requestKeyFrame' }
  | { type: 'participants'; participants: Participant[] }
  | { type: 'remoteTrack'; track: RemoteTrack }
  | { type: 'trackGone'; trackGone: RemoteTrackID }
  | { type: 'update'; update: UpdateInfo }
  | { type: 'log'; log: LogEntry }
  | { type: 'metrics'; metrics: Metrics }
  | { type: 'error'; error: string };

/** Endpoint descriptor served by the backend at /__bridge. */
export interface Endpoint {
  url: string;
  token: string;
  /** The build the backend — and so this frontend — was packaged as. */
  version?: string;
}

/**
 * Normalises whatever WebCodecs hands back as a codec description into a
 * Uint8Array. The spec types `description` as AllowSharedBufferSource, so
 * it can be an ArrayBuffer, a SharedArrayBuffer, or any view over either.
 */
export function toBytes(src: AllowSharedBufferSource): Uint8Array {
  if (src instanceof ArrayBuffer) return new Uint8Array(src);
  const view = src as ArrayBufferView;
  return new Uint8Array(view.buffer as ArrayBuffer, view.byteOffset, view.byteLength);
}

/** base64-encodes raw bytes for the JSON control channel. */
export function toBase64(bytes: Uint8Array): string {
  let s = '';
  for (let i = 0; i < bytes.length; i++) s += String.fromCharCode(bytes[i]);
  return btoa(s);
}

/** Decodes a base64 codec description back to bytes. */
export function fromBase64(b64: string): Uint8Array {
  const s = atob(b64);
  const out = new Uint8Array(s.length);
  for (let i = 0; i < s.length; i++) out[i] = s.charCodeAt(i);
  return out;
}
